### PR TITLE
feat: add /servers for user server listing

### LINKED ISSUE #125: Add repository-specific metadata to pod metadata
Use pod annotations to connect the pod with 

- namespace
- project
- commit-sha

### LINKED ISSUE #127: Modify the user server list to include namespace/project/commit-sha
Add pod annotations (#125) to user server list returned from the `/jupyterhub/services/notebooks/user` endpoint.

### PR DESCRIPTION
**Pull request by rokroskar**
Adds project-specific metadata (namespace/project/branch/commit)
to the servers json.

closes #125 
closes #127

### CODE REVIEW

File: src/notebooks_service.py [hunk 1]
```diff
@@ -196,8 +221,34 @@ def ui(path):
...
+            pod.metadata.name: pod.metadata.annotations
+            for pod in pods
+        }
+
+        for server_name, properties in servers.items():
+            pod_annotations = annotations.get(
+                properties.get('state', {}).get('pod_name', {})
+            )
+            properties[RENKU_ANNOTATION_PREFIX] = {
+                key[len(RENKU_ANNOTATION_PREFIX+'/'):]: value
```
Comment (jirikuncar) [inline src/notebooks_service.py]: I would set the annotations as `servers[server_name]['annotations'] = {key: value for (key, value) in pod_annotations.items() if key.startswith(RENKU_ANNOTATION_PREFIX)}`

Comment (rokroskar) [reply src/notebooks_service.py]: 👍 for calling it `annotations` - but I would remove the `RENKU_ANNOTATION_PREFIX` from the key anyway, it's not really necessary is it?

Comment (jirikuncar) [reply src/notebooks_service.py]: it might be nice if we decide to allow other prefixes in the future.

**Comment by lorenzo-cavazzi** [comment]
is this addressing #125 ? The additional data are added to `jupyterhub/services/notebooks/servers` but not to `jupyterhub/services/notebooks/user`

**Comment by rokroskar** [comment]
I think the `user` endpoint could go away - I believe the only thing used from there is `servers`, no?

**Comment by lorenzo-cavazzi** [comment]
> I think the `user` endpoint could go away - I believe the only thing used from there is `servers`, no?

Yes, as far as I know `user` endpoint is currently used only by notebook server tab. If it's ok for you we can leave it there as it is and remove it as soon as I update the UI

**Comment by rokroskar** [comment]
Sure, sounds good

File: src/notebooks_service.py (hunk 1 above)
Comment (rokroskar) [reply src/notebooks_service.py]: ok I fixed it up, see the sample JSON below

**Comment by rokroskar** [comment]
Here's the sample JSON as it currently stands:

```
{
  "servers": {
    "rokroskar-jumping-2d-ed3e084": {
      "annotations]": {
        "renku.io/branch": "master", 
        "renku.io/commit-sha": "ed3e0844289b8e6996543b1c838fed4aa6ffe07c", 
        "renku.io/namespace": "rokroskar", 
        "renku.io/project": "jumping-junipers"
      }, 
      "last_activity": "2019-04-02T09:49:26.008000Z", 
      "name": "rokroskar-jumping-2d-ed3e084", 
      "pending": null, 
      "progress_url": "/jupyterhub/hub/api/users/rokroskar/servers/rokroskar-jumping-2d-ed3e084/progress", 
      "ready": true, 
      "started": "2019-04-02T09:49:02.720858Z", 
      "state": {
        "pod_name": "jupyter-rokroskar-rokroskar-2djumping-2d2d-2ded3e084"
      }, 
      "url": "/jupyterhub/user/rokroskar/rokroskar-jumping-2d-ed3e084/"
    }
  }
}
```

**Comment by lorenzo-cavazzi** [comment]
`"annotations]": {` is correct? Or should if be`"annotations": {` ?

**Comment by rokroskar** [comment]
arghhhhhhhhhhhhhh thanks

**Comment by rokroskar** [comment]
@lorenzo-cavazzi ok should be fixed in b6cf96c

File: jupyterhub/spawners.py [hunk 2]
```diff
@@ -343,13 +354,22 @@ def get_pod_manifest(self):
...
-            # Set the notebook container image
-            self.image_spec = self.image
-
             # Set the repository path to the working directory
             self.working_dir = mount_path
             self.notebook_dir = mount_path
 
+            # add git project-specific annotations
+            self.extra_annotations = {
+                RENKU_ANNOTATION_PREFIX + '/namespace':
```
Comment (jirikuncar) [inline jupyterhub/spawners.py]: maybe the prefix should contain the `/` to avoid matching `renku.iot`

File: src/notebooks_service.py [hunk 3]
```diff
@@ -196,8 +221,33 @@ def ui(path):
...
+        }
+
+        for server_name, properties in servers.items():
+            pod_annotations = annotations.get(
+                properties.get('state', {}).get('pod_name', {})
+            )
+            servers[server_name]['annotations'] = {
+                key: value
+                for (key, value) in pod_annotations.items()
+                if key.startswith(RENKU_ANNOTATION_PREFIX)
```
Comment (jirikuncar) [inline src/notebooks_service.py]:
```suggestion
                if key.startswith(RENKU_ANNOTATION_PREFIX + '/')
```

File: jupyterhub/spawners.py (hunk 2 above)
Comment (rokroskar) [reply jupyterhub/spawners.py]: good point -- I'll make the change in the next PR

**Pull request merged**